src/Domain/ZCurve.hpp

```diff
@@ -23,7 +25,52 @@ namespace domain {
 /// v  1 |  1   3   5   7
 /// \endcode
 ///
+/// For the Z-curve mapping, this function takes you from `ElementId` to
+/// Z-curve index, whereas `domain::segment_indices_from_z_curve_index` is the
+/// inverse and takes you from Z-curve index to segment indices.
+///
 /// \param element_id the `ElementId` for which to compute the Z-curve index
 template <size_t Dim>
 size_t z_curve_index_from_element_id(const ElementId<Dim>& element_id);
+
+/// \brief Computes the `Segment` indices of an `Element` from a given Z-curve
+/// index and block refinements
+///
+/// \details The `Segment` indices are computed by extracting the bits of the
+/// `Segment` indices from the Z-curve index. Here is a sketch of a 2D block
+/// with 4x2 elements, with bit indices and the resulting z-curve:
+///
+/// \code
+///        x-->
+///        00  01  10  11
+/// y  0 |  0   2   4   6
+/// |    |
+/// v  1 |  1   3   5   7
+/// \endcode
+///
+/// For the Z-curve mapping, this function takes you from Z-curve index to
+/// segment indices, whereas `domain::z_curve_index_from_element_id` is the
+/// inverse and takes you from `ElementId` to Z-curve index.
+///
+/// \param z_curve_index the Z-curve index for which to compute the `Segment`
+/// indices
+/// \param initial_ref_levs the refinements of the block that the `ElementId`
+/// belongs to
+template <size_t Dim>
+std::array<size_t, Dim> segment_indices_from_z_curve_index(
+    size_t z_curve_index, const std::array<size_t, Dim>& initial_ref_levs);
+
+/// \brief Create the `ElementId`s of a single `Block` ordered by their Z-curve
+/// index
+///
+/// \details For details on `ElementId`'s Z-curve indices, see
+/// `domain::segment_indices_from_z_curve_index`
+///
+/// \param block_id the `Block` number
+/// \param initial_ref_levs the refinements levels of the `Block`
+/// \param grid_index the grid index of the `Block`
+template <size_t Dim>
+std::vector<ElementId<Dim>> initial_element_ids_in_z_curve_order(
```

Comment:
That should work! I only defined the inverse to avoid sorting, but if sorting is the preference, I can do that and yes, it would reduce the amount and simplify the code

Reply:
🤷‍♂️ I don't know, sorting is a lot easier and probably not prohibitively expensive, but I haven't thought much about this. Definitely keep the inverse code around on a branch if you decide to go the sorting route. Also maybe other reviewers have a more educated opinion here?

Reply:
The implementation is preserved on other branches, regardless of what we choose. I can wait to see what others say!

Reply:
I haven't thought about this particular problem, but the general advice I've seen with respect to sorting is that sorting is fast.  (Although sorting something that is nearly in order is not faster.)  In the asymptotic regime, which is probably actually relevant here, it's `O(n log n)`, and if you have a vector you were probably planning to do something at least `O(n)` anyway.  Of course if the comparator is expensive that could be an issue, but it doesn't look too bad here (although it probably adds another factor of `O(log n)`, which is fairly irrelevant).

The only question then is if the overall prefactor gets a lot worse.  Here, the inverse function being used in the current algorithm looks about as complex as the function used in the comparator, so it's probably not an issue.  If it is, there's an easy factor of ~2 to be gained by precomputing the indices of all the elements.